src/net.cpp

```diff
@@ -340,71 +340,161 @@ bool CConnman::CheckIncomingNonce(uint64_t nonce)
+            FindNode(addrConnect.ToStringIPPort()))
+            return false;
+    } else if (FindNode(std::string(pszDest))) {
+        return false;
     }
 
     /// debug print
     LogPrint(BCLog::NET, "trying connection %s lastseen=%.1fhrs\n",
         pszDest ? pszDest : addrConnect.ToString(),
         pszDest ? 0.0 : (double)(GetAdjustedTime() - addrConnect.nTime)/3600.0);
 
-    // Connect
-    SOCKET hSocket;
-    bool proxyConnectionFailed = false;
-    if (pszDest ? ConnectSocketByName(addrConnect, hSocket, pszDest, Params().GetDefaultPort(), nConnectTimeout, &proxyConnectionFailed) :
-                  ConnectSocket(addrConnect, hSocket, nConnectTimeout, &proxyConnectionFailed))
-    {
-        if (!IsSelectableSocket(hSocket)) {
-            LogPrintf("Cannot create connection: non-selectable socket created (fd >= FD_SETSIZE ?)\n");
-            CloseSocket(hSocket);
-            return NULL;
+    std::string strDest;
+    int port = Params().GetDefaultPort();
+    CAddress addrDest;
+
+    // Check input.
+    // If pszDest is set:
+    // - It must contain parseable host, and optionally a port.
+    // - If no port is supplied, the default is used.
+    // Otherwise, addrDest must be valid.
+    if (pszDest) {
+        SplitHostPort(std::string(pszDest), port, strDest);
+        if (strDest.empty()) {
+            return false;
+        }
+    } else {
+        addrDest = addrConnect;
+        if (!addrDest.IsValid()) {
+            return false;
         }
+    }
 
-        if (pszDest && addrConnect.IsValid()) {
-            // It is possible that we already have a connection to the IP/port pszDest resolved to.
-            // In that case, drop the connection that was just created, and return the existing CNode instead.
-            // Also store the name we used to connect in that CNode, so that future FindNode() calls to that
-            // name catch this early.
-            LOCK(cs_vNodes);
-            CNode* pnode = FindNode((CService)addrConnect);
-            if (pnode)
-            {
-                pnode->MaybeSetAddrName(std::string(pszDest));
-                CloseSocket(hSocket);
-                LogPrintf("Failed to open new connection, already connected\n");
-                return NULL;
+    if (!strDest.empty()) {
+        // Resolve strDest, if set. Only actually perform a dns lookup if they're enabled
+        // by the current settings. Note that if strDest is already an ip address, this will
+        // succeed regardless of the lookup setting.
+        std::vector<CService> addrResolved;
+        if (Lookup(strDest.c_str(), addrResolved, port, fNameLookup && !HaveNameProxy(), 256)) {
+            if (addrResolved.size() > 0) {
+                addrDest = CAddress(addrResolved[GetRand(addrResolved.size())], NODE_NONE);
             }
         }
+    }
 
-        addrman.Attempt(addrConnect, fCountFailure);
+    if (!strDest.empty() && addrDest.IsValid()) {
```

Comment:
While it may make sense to lump a few actions together if they meet the same conditions now because a connection is just one long synchronous operation, my intention is to clarify which conditions really need to be met at which stage, so that we can break that operation up into smaller asynchronous ones.

Some of these tests (like whether or not we resolved successfully) will just move in an obvious way to callbacks, so I guess it makes sense to combine them here. Others though, will need to be re-checked on separate threads.

I'll clarify this in the code by setting some variables rather than repeating conditions, so that it's more clear what the intent is.

I'll rebase while I'm at it.